### PR TITLE
SocketsHttpHandler: do not use DualMode sockets in default connection logic

### LINKED ISSUE #44686: Azure App Service HTTP requests to Azure VNet IP Addresses fail after upgrading to .NET 5.0
I have an ASP.NET Core 3.1 application running in Azure App Service. It makes calls to Elasticsearch running on an Ubuntu Server 16.04 LTS Azure Virtual Machine, connected to the App Service via an Azure VNet. When I upgraded the application to target `net5.0`, all of my calls to Elasticsearch started failing with the following exception:

```
System.Net.Http.HttpRequestException: An attempt was made to access a socket in a way forbidden by its access permissions. (10.1.0.5:9200)
 ---> System.Net.Sockets.SocketException (10013): An attempt was made to access a socket in a way forbidden by its access permissions.
```

Through trial and error, I eliminated Elasticsearch from the equation and reproduced the issue with simple HTTP requests via `HttpClient`:

```csharp
public async Task<IActionResult> VNetRequest()
{
    string? body;
    try
    {
        using var requestMsg = new HttpRequestMessage(HttpMethod.Get, _testRequest.VNetEndpoint);
        using var responseMsg = await _httpClient.SendAsync(requestMsg);
        body = await responseMsg.Content.ReadAsStringAsync();
    }
    catch (Exception ex)
    {
        body = ex.ToString();
    }

    return Content(body);
}
```

When targeting `netcoreapp3.1`, this call will succeed and display some nginx boilerplate text.

When targeting `net5.0`, this call will fail with the following exception:

```
System.Net.Http.HttpRequestException: An attempt was made to access a socket in a way forbidden by its access permissions. (10.1.0.4:80)
 ---> System.Net.Sockets.SocketException (10013): An attempt was made to access a socket in a way forbidden by its access permissions.
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.ThrowException(SocketError error, CancellationToken cancellationToken)
   at System.Net.Sockets.Socket.AwaitableSocketAsyncEventArgs.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at System.Net.Http.HttpConnectionPool.DefaultConnectAsync(SocketsHttpConnectionContext context, CancellationToken cancellationToken)
   at System.Net.Http.ConnectHelper.ConnectAsync(Func`3 callback, DnsEndPoint endPoint, HttpRequestMessage requestMessage, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at System.Net.Http.ConnectHelper.ConnectAsync(Func`3 callback, DnsEndPoint endPoint, HttpRequestMessage requestMessage, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.ConnectAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.CreateHttp11ConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.GetHttpConnectionAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean async, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at System.Net.Http.DiagnosticsHandler.SendAsyncCore(HttpRequestMessage request, Boolean async, CancellationToken cancellationToken)
   at Microsoft.Extensions.Http.Logging.LoggingHttpMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at Microsoft.Extensions.Http.Logging.LoggingScopeHttpMessageHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.SendAsyncCore(HttpRequestMessage request, HttpCompletionOption completionOption, Boolean async, Boolean emitTelemetryStartStop, CancellationToken cancellationToken)
   at AppServiceVNetCalls.Controllers.HomeController.VNetRequest() in D:\home\site\repository\src\AppServiceVNetCalls\Controllers\HomeController.cs:line 34
```

When I run these tests locally, I don't get any failures; everything works fine. 

At this point, I'm stuck, and I don't know how where to go next.

# Reproduction repository and Azure applications #

I have a simple [ASP.NET Core application that demonstrates the issue](https://github.com/jonsagara/AppServiceVNetCalls/tree/net5). This repository has two branches, `net31` and `net5`:

- `net31`: This demonstrates that `HttpClient` GET requests to an Azure VNet IP address succeed. It also demonstrates that requests to a non-Azure VNet IP address succeed. See: https://appservicevnettest31.azurewebsites.net/
- `net5`: This demonstrates that `HttpClient` GET requests to an Azure VNet IP address *fail*. It also demonstrates that requests to a non-Azure VNet IP address succeed. See: https://appservicevnettest50.azurewebsites.net/

The architecture of my Azure reproduction is as follows:

- An Azure App Service hosting the ASP.NET applications. It's an S1 App Service Plan hosted in US West. They are Windows plans.
- An Azure Virtual Machine running Ubuntu Server 20.04 LTS on a Gen2 Standard B1s VM. I installed nginx solely to respond on port 80.
- The VM is connected to the App Service via an Azure VNet. As I am not a systems engineer, setting this up was laborious, and I'm not sure I could properly document it. However, it did require setting up an Azure Gateway VPN to connect the App Service to the VM.

Thank you.

ETA: Please let me know when you no longer need these VMs/App Services so that I can delete them. Thanks.

### PR DESCRIPTION
**Pull request by antonfirsov**
In some environments IPV6 (thus dual-stack) sockets do not work despite `Socket.SupportsIPv6` returning true.

I'm reverting the default branch of `HttpConnectionPool.ConnectToTcpHostAsync` to use the logic in 628d99b624f338b0eaca031adc50952b4c2d509d (before #39524) to fix #44686.

The only thing I changed in the old code is comments.

@scalablecory @geoffkizer PTAL.

### CODE REVIEW

**Review by scalablecory — APPROVED** [review]

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs
Comment (gfoidl) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:62]: Nit:
```suggestion
            catch (Exception error) when (error is not OperationCanceledException)
```
Reads a bit nicer.

**Comment by stephentoub** [comment]
I don't understand something.  If the problem is that the OS doesn't allow dual-stack sockets, why isn't Socket.DualMode throwing an exception when we try to set it?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs
Comment (stephentoub) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:25]: This is all concerning to me.  The premise of the SocketsHttpHandler.ConnectCallback's design (and not exposing the default implementation of how a socket is created or the default connect logic) was that it's just a couple of lines of code to emulate the default behavior... now it's over a 100 lines?
cc: @geoffkizer

**Comment by scalablecory** [comment]
> I don't understand something. If the problem is that the OS doesn't allow dual-stack sockets, why isn't Socket.DualMode throwing an exception when we try to set it?

It seems like the OS as a whole does, but the NIC driver or some firewall setting is blocking it at a later point.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs
Comment (scalablecory) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:25]: We're going to introduce an easy mode static `Socket.ConnectAsync` API, so it'll go back to a 5 liner.

Comment (stephentoub) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:25]: Then shouldn't we add that and then use it here rather than doing it this way? I don't see why we're putting back this connect helper logic.  If the concern is being able to more easily backport, that same static API can just be added as an internal SocketEx static in the backport or something like that.

**Comment by stephentoub** [comment]
So our message then is don't use DualMode because something somewhere may not support it even if the system says it does?

**Comment by wfurt** [comment]
From the conversion, it seems like the Azure VPN ignores the dual-mode socket and the packet is than routed to physical interface instead of private VPN and then rejected by firewall (or lost as it missed tunnel) 
It seems like 3.1 worked by pure luck IMHO -> there is nothing really wrong with 5.0. The fundamental flaw lives in the Azure implementation as far as I can tell.

**Comment by stephentoub** [comment]
@wfurt, that's what it sounded like to me,  too, which is why I'm skeptical of the direction of this change.

**Comment by davidfowl** [comment]
And AWS

**Comment by antonfirsov** [comment]
@stephentoub people are also hitting it in AWS:
 https://twitter.com/bhop2112/status/1334943693179117570

So far I'd say there is a new user confirming the issue every week. Do we want to tell each of them to use the workaround? For some it's not possible, because they are using `HttpClient` through external libraries (2 users reported with Elastic Search).

**Comment by scalablecory** [comment]
I agree it seems like dual-mode should work fine here, and these environments are doing something wrong.

**Comment by stephentoub** [comment]
> Do we want to tell each of them to use the workaround?

If the environments are broken,  we should work with the powers that be to fix the environments.  If alternatively we're saying dual-mode should never be used because it can't be trusted,  we should be explicit about that, find and fix all such usage in all of our code,  deprecate the relevant ctors/properties, etc etc.

**Comment by antonfirsov** [comment]
@stephentoub I don't think we are saying that dual-mode should be never used. In this solution we are switching to a default connect implementation that provides the highest compatibility counting with flawed environments. Users normally do not need to make such compatibility efforts in their ConnectCallbacks.

**Comment by antonfirsov** [comment]
I don't have a strong opinion here, I think I raised more or less the same concern in my https://github.com/dotnet/runtime/issues/44686#issuecomment-738189574 as @stephentoub here in the PR, but @scalablecory @geoffkizer and @karelz voted for the trivial workaround. Now it seems opinions are changing.

There are basically 3 options:
1. Do not "solve" this in .NET, keep using dual-stack. While waiting for the solution from the cloud providers, communicate the [workaround](https://github.com/dotnet/runtime/issues/44686#issuecomment-733797994), and leave the users who do not have control over `HttpClient` without help. Note that we don't have an ETA from Azure and no working relationship with AWS AFAIK.
2. Go with the cheapest possible solution (this PR), but probably communicate dual-stack sample code in docs as suggested by @scalablecory .
3. Something more sophisticated, that [involves a new API](https://github.com/dotnet/runtime/issues/44686#issuecomment-738270840) in .NET 6, and an internal "SocketEx" utility in the backport. Obviously, this is gonna take longer, which is up to @karelz to prioritize then.

How do we plan to decide?

**Comment by stephentoub** [comment]
So when someone uses the Socket(SocketType, ProtocolType) ctor, directly or via some other library that does so,  we do or do not expect that to work in their favorite cloud?

**Comment by antonfirsov** [comment]
In my opinion the `Socket(SocketType, ProtocolType)` constructor is a very bad API, since it's not trivial from the API shape that it's going to create a dual-stack socket, which is still not (fully) supported in a certain environments. I've seen shooting ourselves in the foot several times because of that even in our own code & tests. I would have created something like `Socket.CreateDualMode(SocketType, ProtocolType)` for that instead.

But I think we are mixing concerns in this discussions now:
- How mature is dual-stack / IPV6 adaption? -- Users who interface `Socket` directly, usually know their environment, and have the answers. I don't think we should obsolete fundamental API-s just because they don't work on certain flawed environments.
- The problem comes when the fact whether IPv6 works or not becomes an implementation detail, like with `HttpClient`. We need to decide whether we want to provide a compatibility workaround or not.

**Comment by scalablecory** [comment]
To be clear, while I think this is an environment problem, I still feel we should make this change for compat purposes.

I do not think we should recommend against dual-mode, but I don't mind us making something a little bit more compatible by default when customers using common cloud environments are affected.

**Comment by scalablecory** [comment]
(That said, we also need to reach out to these platforms to understand why they do it this way, and ask them to fix it if there is no why and it's just a config problem)

**Comment by stephentoub** [comment]
I'm pushing on this for a few reasons.

This isn't just about HttpClient.  Yes, it's a key library, but there are others, and if `new Socket(SocketType, ProtocolType)` doesn't work in the clouds devs deploy their apps to, and if we're pushing developers to this (which we are, if for no other reason than it's the simplest ctor for this purpose, shows up first in IntelliSense because it's shorter than the three argument version, etc.), then we're setting ourselves up for failure as a cloud native stack that doesn't work in the cloud.  It's easy to find other libraries relying on this as well, for example:
- SignalR: https://github.com/aspnet/SignalR/blob/8c6ed160d84f58ce8edf06a1c74221ddc8983ee9/samples/ClientSample/Tcp/TcpConnection.cs#L26
- DotNetty: https://github.com/Azure/DotNetty/blob/47f5ec7303037f1360615d182939e04d8619a2b3/src/DotNetty.Transport/Channels/Sockets/TcpServerSocketChannel.cs#L30-L31
- Akka.net: https://github.com/akkadotnet/akka.net/blob/a474a886bf0c8944760724f01fafb13cf6b8e404/src/core/Akka/IO/Inet.cs#L80
- The socks proxy we tell devs about: https://github.com/MihaZupan/HttpToSocks5Proxy/blob/f595aa19b000025ee53081b8607db29c26740afa/src/HttpToSocks5Proxy/HttpToSocks5Proxy.cs#L403-L404
Are all of these broken if deployed to Azure or AWS?  That sounds really strange, but if it's true, that's really bad.  And if it's not true, then our understanding of the cause seems to have some gaps that we should fill in before pushing for a solution.

With regards to HttpClient, we've pushed the ConnectCallback as the thing that addresses everyone's woes.  Need to bind to a particular interface? Use ConnectCallback.  Need to configure TTL?  Use ConnectCallback.  Need to set Receive/SendTimeout?  Use ConnectCallback.  Etc.  It's going to be used, and not infrequently.  Which is a good thing, except that the solution we've put forth now apparently has impliciations we didn't previously understand.  For the next year, developers that follow-suit with what we've done in .NET 5 are either going to a) have to write more than a hundred lines of code to get the "right" behavior, or b) suffer the same "this no longer works in the preeminent clouds" behavior, assuming that's the actual impact.  That's not good.  Even if we introduce a new `Socket.ConnectAsync(...)` static that puts this back to the simpler solution, that won't ship for another year.  What is our recommendation until then?

I want to make sure we have our understanding and story straight before we rush in something "for compatibility", because while that may end up being the right answer, it has consequences. At the end of the day, the right answer may be to merge a fix into release/5.0 that reverts to the more complicated scheme, but let's _really_ understand what's going on, having spoken with the folks at Azure and AWS, before we do so, as it has ramifications for lots of other things.  And if it turns out that libraries and apps using `new Socket(SocketType, ProtocolType)` really are doomed for use in the cloud, then how can we not deprecate it?  Who would we recommend use it at that point?

With regards to the actual fix, @scalablecory mentioned exposing a static `Socket.ConnectAsync` as the actual plan, but I don't see that API called out in https://github.com/dotnet/runtime/issues/43935.  Did I miss it?  With regards to the actual fix, this PR is targeting .NET 6, but if I'm understanding correctly, this is not what we'd intend to ship in .NET 6.  Rather, we'd intend to add a that "simple" static, and the change in this PR to System.Net.Http would be limited to tweaking this:
https://github.com/dotnet/runtime/blob/dc0e11d2e681659fc20feebf4c092082e75706af/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs#L1301-L1316
to something like this:
```C#
                    // Otherwise, create and connect a socket using default settings.
                    if (async)
                    {
                        socket = await Socket.ConnectAsync(endPoint, cancellationToken).ConfigureAwait(false);
                    }
                    else
                    {
                        socket = new Socket(SocketType.Stream, ProtocolType.Tcp);
                        using (cancellationToken.UnsafeRegister(static s => ((Socket)s!).Dispose(), socket))
                        {
                            socket.Connect(endPoint);
                        }
                    }

                    socket.NoDelay = true;
                    return new NetworkStream(socket, ownsSocket: true);
```
yes?  So, is the thinking the "simpler" solution is churning all this code in master, porting that back to release/5.0, and then immediately churning master again to the real .NET 6 solution?

**Comment by antonfirsov** [comment]
@stephentoub the API is indeed missing, we need to propose it: https://github.com/dotnet/runtime/issues/44686#issuecomment-738270840

> having spoken with the folks at Azure and AWS

Just added you to the email thread we have with Azure about this. The problem is understood, solution seems to be on the way, but there is no ETA. I have no idea who and how could connect AWS on this matter.

**Comment by scalablecory** [comment]
@normj can you help us with AWS side of this?

**Comment by antonfirsov** [comment]
Context on AWS: https://github.com/dotnet/runtime/issues/44686#issuecomment-738994043

**Comment by davidfowl** [comment]
We had similar problems with Bind in kestrel (it binds to both ipv4 and ipv6 by default) and we fallback to ip4v of binding fails.

Cc @Tratcher @halter73

**Comment by geoffkizer** [comment]
> In my opinion the Socket(SocketType, ProtocolType) constructor is a very bad API, since it's not trivial from the API shape that (typically) it's going to create a dual-stack socket, which is still not (fully) supported in a certain environments. I've seen shooting ourselves in the foot several times because of that even in our own code & tests. I would have created something like Socket.CreateDualMode(SocketType, ProtocolType) for that instead.

Totally agree. Doesn't help with the issue at hand, but something to consider for the future.

**Comment by geoffkizer** [comment]
@stephentoub 

> If the environments are broken, we should work with the powers that be to fix the environments. If alternatively we're saying dual-mode should never be used because it can't be trusted, we should be explicit about that, 

I think we are saying both. First, the environments are broken and we should push them to fix this. But second, since this is how they behave today, you should never use dual-mode in these environments (or in code that could run in these environments) because it can't be trusted today.

> find and fix all such usage in all of our code, 

Yes -- given what we know about these environments, I don't think we should ever use dual mode sockets in our own code.

> deprecate the relevant ctors/properties, etc etc.

I think that's going a little far. If you know your target environment supports dual-mode sockets, then feel free to use dual-mode sockets.

That said, as @antonfirsov pointed out above, the structure of the API makes it such that it's often not obvious that you are using dual-mode sockets. So perhaps we should do something to address that.

**Comment by geoffkizer** [comment]
@stephentoub 

> Are all of these broken if deployed to Azure or AWS? That sounds really strange, but if it's true, that's really bad.

I think the answer is: All of them are broken when deployed to Azure or AWS *in certain deployment configurations*. 

What I don't have a good sense of is how common these configurations are. It would be good to get more data here.

But regardless, I think we have enough data to say that they are common enough to justify changing HttpClient to not use dual-mode sockets, and probably for other libraries as well, depending on how they are deployed.

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:76]: Isn't this creating a dual-mode socket?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs
Comment (geoffkizer) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:25]: I agree with @stephentoub here

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs
Comment (geoffkizer) [inline src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:1284]: The old try catch wrapped both the ConnectCallback case as well as the default case. This one only wraps the ConnectCallback case. Why the change?

**Comment by geoffkizer** [comment]
> It's going to be used, and not infrequently. Which is a good thing, except that the solution we've put forth now apparently has impliciations we didn't previously understand. For the next year, developers that follow-suit with what we've done in .NET 5 are either going to a) have to write more than a hundred lines of complicated code to get the "right" behavior, or b) suffer the same "this no longer works in the preeminent clouds" behavior, assuming that's the actual impact. 

I agree with all of this, but I'm not sure what to do about it.

I think we should be doing all of the following:
- Push Azure/AWS/etc to support dual-mode sockets
- Add a Task-based static Socket.ConnectAsync for 6.0, and use it in HttpClient
- Backport that to 5.0 using private APIs so we aren't exposing new API

But unless we are willing to expose new API in 5.0, I think we are kinda stuck in a not ideal place here. Thoughts?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs
Comment (antonfirsov) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:76]: Yes, but there was no sync in 3.1. I dont like this inconsistence either, just took the state at the commit I mentioned.

The bad thing is that the ony alternatives I see is to either do sync over async or implement a sync version of [DnsConnect](https://github.com/dotnet/runtime/blob/34cae7461f96a9c1a7574c72c436e309bf4031f4/src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.cs#L655) within System.Net.Http

**Comment by antonfirsov** [comment]
Since it's [our IPv6 support detection logic](https://github.com/dotnet/runtime/blob/34cae7461f96a9c1a7574c72c436e309bf4031f4/src/libraries/Common/src/System/Net/SocketProtocolSupportPal.Windows.cs#L24-L30) that is unreliable in these environments, I think we should also consider an environment variable to force-disable IPv6 as an alternative to this PR.

Pros: Not specific to HttpClient, fixes the problem for all users of `Socket(SocketType, ProtocolType)`. Simple and quick solution. Communicates the fact that the problem is with the environment.

Cons: not automatic, .NET 5 update is still a breaking change that has to be actioned by the customers.

**Comment by stephentoub** [comment]
> Since it's our IPv6 support detection logic that is unreliable in these environments

Can we improve that logic? For example, right now we only try to create the socket. What happens if we try to bind it? What happens if we try to connect it (over loopback)? Etc.

**Comment by antonfirsov** [comment]
@stephentoub I would be happy to experiment with all of this ... if I had access to a repro environment, or if I knew how to create one.

There were some emails about in the past 12hours, we are working under time pressure now.

**Comment by ManickaP** [comment]
If the default connect logic will be whole of this:
https://github.com/dotnet/runtime/blob/6cba345ff9bcf2b55d27b64f0d5e08398eb10cd7/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs#L25-L70

How are we going to communicate to people how to implement their own `ConnectCallback` if we're not providing this implementation on the outside and it's this long? I can imagine that some of the callback implementations will just want to enhance the logic, not completely replace it. With this, they'll have to copy paste 50 lines of code or reach for the method via reflection.

I understand this needs to be solved for 5.0 and rather quickly, but can't we at least think of something more user-friendly before we ship this? Also, isn't the same problem happening for the sync code path?

**I really like the idea of solving this at the socket level.**

**Comment by jonsagara** [comment]
> I would be happy to experiment with all of this ... if I had access to a repro environment, or if I knew how to create one.

@antonfirsov  Would it help if I gave you commit access to my repro repository? I think I'd need to change the deployment from `kudu` to a self-contained application via Azure DevOps so that you can deploy. `kudu` only supports specific, pre-intalled versions of .NET.  As long as your new test version of .NET 5 is available to be downloaded as a preview, I think I could make this work with a simple DevOps deployment.

Or, maybe I could give you direct publish access to the App Service so that you can push your own self-contained app directly?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs
Comment (antonfirsov) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs:1284]: There is an equivalent catch block in `ConnectHelper.ConnectAsync`:

https://github.com/dotnet/runtime/blob/6cba345ff9bcf2b55d27b64f0d5e08398eb10cd7/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs#L62-L65

https://github.com/dotnet/runtime/blob/6cba345ff9bcf2b55d27b64f0d5e08398eb10cd7/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs#L239-L244

**Comment by antonfirsov** [comment]
@ManickaP @jonsagara not decided yet, but we may ship a quickfix for the January update of .NET 5. If we do so we need to act .. very quickly.

If we want to catch that train, we need to merge either this PR or #45893 as is. We can then continue investigations about solving this at socket level either by making IPv6 detection more robust or by shipping & backporting new static `Socket.ConnectAsync` and `Socket.Connect` overloads.

**Comment by antonfirsov** [comment]
> Would it help if I gave you commit access to my repro repository?

@jonsagara does pushing trigger deployment?

File: src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs
Comment (antonfirsov) [reply src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectHelper.cs:62]: I prefer to not do refactors in this PR, there are way too many thinks we may want to fix.

**Comment by jonsagara** [comment]
> @jonsagara does pushing trigger deployment?

It does, but I don't think that will work for this case. App Service `kudu` deployment currently only supports .NET SDK 5.0.100.

We can get around this by directly publishing your changes as a self-contained deployment. All I'd need to do is give you the Publish Profile from Azure Portal. Then, clone the repo, import the Publish Profile, and you should be able to publish and test your changes as a self-contained application.

I just tested this using a currently unsupported build (.NET SDK 5.0.101), and the app started and still reproduces the bug.

I can't share the Publish Profile publicly, so if you're interested, please let me know the best place to send it. Thanks!

**Comment by antonfirsov** [comment]
@jonsagara DM me on twitter: https://twitter.com/antonfrv

**Comment by jonsagara** [comment]
@antonfirsov Will do. Will you please follow me back so that I can send you a message? https://twitter.com/jonsagara

**Comment by karelz** [comment]
Closing the PR as we do not plan to fix the problem this way anymore -- see https://github.com/dotnet/runtime/issues/44686#issuecomment-745467841 for details.

**Pull request closed without merging**
